---
title: "'Other Omics' practical"
author: "Matthew Suderman"
output: 
  html_document:
    theme: flatly
    number_sections: no
    df_print: paged
    code_folding: show
---

```{css, echo = FALSE}

blockquote {
    padding: 10px 20px;
    margin: 0 0 20px;
    font-size: 16px;
    border-left: 5px solid #eee;
}

```

Set `show.answers` to `TRUE` to reveal answers to questions
in this practical.

```{r setup, include=FALSE}
knitr::opts_chunk$set(message=FALSE,warning=FALSE)
knitr::read_chunk('answers.r')
show.answers <- F 
```

### Objectives

-   Prepare the data for analysis.
-   Conduct an eQTL analysis.
-   Produce plots to visually inspect findings.
-   (Optional) Conduct an meQTL analysis.
-   (Optional) Identify trios of associated genes, CpG sites and SNPs.

### Data

In this practical we will run eQTL and meQTL analyses on a small publicly available dataset:

Wagner, et al.
[The relationship between DNA methylation, genetic and expression inter-individual variation in untransformed human fibroblasts.](http://www.ncbi.nlm.nih.gov/pubmed/24555846)
Genome Biol.
2014 Feb 20;15(2):R37.
[GEO link](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE53261)

It includes: 

- Genotype (Illumina Human1M-Duov3 DNA Analysis BeadChip), 

- DNA methylation (Illumina HumanMethylation450 BeadChip), and 

- Gene expression (Illumina HumanRef-8 v3.0 Expression Beadchip)
  profiles from 57 human skin fibroblast samples obtained from Coriell
  and McGill Cellbank.

### Analysis steps

1.  Preprocessing.
    a.  Population stratification.
    b.  Gene expression outliers.
    c.  Matching samples between datasets.
    d.  Non-genetic variation removal.
2.  Software and data.
    a.  R package MatrixEQTL.
    b.  Genetic data.
    c.  Population stratification covariates.
    d.  Expression data.
    e.  Gene and SNP locations.
3.  eQTL analysis.
4.  Aftermath.
    a.  Number of associations.
    b.  Top associations.
    c.  Association plotting.
5.  (Optional) meQTL analysis.
6.  (Optional) Associated trios.

### 1a. Preprocessing: Population stratification

This is detected by visualizing genetic distances between all pairs of
individuals using multi-dimensional scaling (MDS).

We will use PLINK for these computations (https://www.cog-genomics.org/plink/). 

First, make the plink binary executable:
```{bash}
chmod +x ./plink
```

Calculate genetic distances:
```{bash,results="hide"}
mkdir output
./plink --bfile data/snp --genome --out output/plink
```

The output `output/plink.genome` file 
directory gives all pairwise distances (see the `DST` column).

You can type the following command to see the first few lines in the file:
```{bash}
head output/plink.genome
```

### 1a. Preprocessing: Population stratification in 2-dimensions

We represent these distances in two dimensions as follows:
```{bash,results="hide"}
./plink --bfile data/snp \
		--read-genome output/plink.genome \
		--cluster --mds-plot 2 \
		--out output/plink
```

This generates a file `output/plink.mds` that looks like this:
```{bash}
head expected-outputs/plink.mds
```

* FID - Family ID; 
* IID - Individual ID (or sample name); 
* SOL - Assigned solution code; 
* C1 - Position on first dimension; 
* C2 - Position on second dimension.

## 1a. Preprocessing: Population stratification visually

We'll send outputs from R to the output directory.
For convenience, we'll save the location as a variable.
```{r}
out.dir <- "output"
```

We now plot the coordinates in `output/plink.mds` in R as follows:
```{r}
snp.mds <- read.table(file.path(out.dir, "plink.mds"), header=T, sep="", stringsAsFactors=F)
plot(snp.mds[,"C1"], snp.mds[,"C2"], main="SNP MDS plot", xlab="first", ylab="second", pch=19)
```

> **Question 1:** Can you identify the two potential outlier samples in the plot (using R code)?
```{r question1,include=show.answers}
```

Given how far these samples differ from the rest, we might actually
consider removing them from the analysis.
In this case, we'll attempt to adjust for these differences.

### 1b. Preprocessing: Gene expression outliers

We could use a similar approach to identify gene expression outliers.
```{r}
rna <- read.csv("data/rna-data.csv", row.names=1)
rna.dist <- 1-abs(cor(rna))
rna.mds <- cmdscale(rna.dist,eig=TRUE, k=2)
plot(rna.mds$points[,1], rna.mds$points[,2],
     main="RNA MDS plot", xlab="first", ylab="second", pch=19)
```

### 1c. Preprocessing: Matching samples between datasets

All too often samples are mislabelled in the lab.  Fortunately, it is
possible to computationally detect mismatches between genetic and gene
expression data using genetic and other omic datasets using the DRAMS
tool: <https://github.com/Yi-Jiang/DRAMS>:

> Jiang Y, Giase G, Grennan K, Shieh AW, Xia Y, Han L, Wang Q, Wei Q,
> Chen R, Liu S, White KP, Chen C, Li B, Liu C. DRAMS: A tool to
> detect and re-align mixed-up samples for integrative studies of
> multi-omics data.
> **PLoS Comput Biol**.
> 2020 Apr 13;16(4):e1007522.
> <https://doi.org/10.1371/journal.pcbi.1007522>.

In this case, we will skip this step, assuming that the study authors
haven't made such errors!

It is possible to do the same for genetic and DNA methylation data,
particularly for data generated using the Illumina HumanMethylation450
BeadChip.  It actually measures a selection of SNPs in addition to the
DNA methylation levels.

### 1d. Preprocessing: Non-genetic variation removal

Gene expression measurements can be influenced by factors other than
genetic variation.  To maximize power to detect associations between
genes and genetic variants, we can remove all non-genetic variation
from the gene expression data.  To save time, we will not do that
here.  In case you're curious, here are the basic steps:

1.  Calculate the principal components of the gene expression data.
2.  Test associations between principal components and genetic variants.
3.  Remove any components with strong genetic associations.
4.  Remove variation of the remaining principal components from the gene expression data.

### 2a. Software and data: R package MatrixEQTL

If `MatrixEQTL` has not already been installed, we can install it as shown below.
```{r, eval=F}
BiocManager::install("MatrixEQTL")
# Respond by typing 'yes' to any questions.
```

We then load the library in R as follows:
```{r}
library(MatrixEQTL)
```

### 2b. Software and data: Genetic data

MatrixEQTL operates on `SlicedData` objects.
These objects provide access to large datasets without using huge
amounts of computer memory.

The genetic data is contained in the file `data/snp-data.csv`.

We load it as follows:
```{r,results="hide"}
snp.data <- SlicedData$new()            ## create a new SlicedData object
snp.data$fileDelimiter <- ","           ## columns in the file are separated by commas
snp.data$fileOmitCharacters <- "NA"     ## missing values are represented by "NA"
snp.data$fileSkipRows <- 1              ## the first column provides row names
snp.data$fileSkipColumns <- 1           ## the first row provides column names
snp.data$fileSliceSize <- 2000          ## only load the data 2000 rows at a time to save memory
snp.data$LoadFile("data/snp-data.csv") 
```

Loading the data takes about half a minute.

### 2c. Software and data: Population stratification covariates

When we test associations between genetic variants, we will need to
adjust for population stratification.  This is encoded by the
multi-dimensional scaling coordinates that we created earlier.

As required by `MatrixEQTL`, we load them as `SlicedData` objects:
```{r}
covariates.data <- SlicedData$new()
covariates.data$initialize(t(snp.mds[,c("C1","C2")]))
```

### 2d. Software and data: Expression data

Next we convert the gene expression data to `SlicedData` objects and
verify that samples match the genotype data samples.
```{r}
rna.data <- SlicedData$new()            ## create a new SlicedData object
rna.data$fileDelimiter <- ","           ## columns in the file are separated by commas
rna.data$fileOmitCharacters <- "NA"     ## missing values are represented by "NA"
rna.data$fileSkipRows <- 1              ## the first column provides row names
rna.data$fileSkipColumns <- 1           ## the first row provides column names
rna.data$fileSliceSize <- 2000          ## only load the data 2000 rows at a time to save memory
rna.data$LoadFile("data/rna-data.csv") 
```

Don't forget to make sure that the gene expression and genetic
datasets have the samples in the same order!
```{r}
stopifnot(all(colnames(snp.data) == colnames(rna.data)))
```

### 2e. Software and data: Gene and SNP locations

> **Question 2:** If we tested associations between all SNP-gene pairs,
> how many tests would we have to perform?
```{r question2,include=show.answers}
```

That's a lot of tests.
To save time, we will only consider *cis* pairs (i.e. SNPs that are with 1 million bases of the gene).
`MatrixEQTL` can do this for us if we provide SNP and gene locations in the genome:
```{r}
snp.loc <- read.csv("data/snp-features.csv", row.names=1)
rna.loc <- read.csv("data/rna-features.csv", row.names=1)
```

### 3. QTL analysis: p-value thresholds

> **Question 3:** `MatrixEQTL` requires a p-value threshold for
> significance. To adjust for the number of tests, we could
> use a Bonferroni-adjusted p-value threshold, i.e. divide 0.05
> by the number of tests to be performed. However, because we only test *cis*
> pairs, the exact number of difficult to estimate. Can you think of
> a way to compute simple lower and upper bounds on the number of tests?
```{r question3,include=show.answers}
```

We'll use the lower bound (178232) to estimate a p-value threshold so that we
don't miss any associations with statistically significant p-values.
```{r}
threshold <- 0.05/178232
```

### 3. eQTL analysis: finally!

Finally we are ready to run the eQTL analysis.
```{r}
eqtl <- Matrix_eQTL_main(
    snps = snp.data,
    gene = rna.data,
    cvrt = covariates.data, ## population stratification
    pvOutputThreshold = 0, ## consider only cis pairs
    pvOutputThreshold.cis = threshold, 
    output_file_name.cis = file.path(out.dir, "eqtl-cis.txt"),
    snpspos = snp.loc, 
    genepos = rna.loc[,c("geneid","chr","left","right")],
    cisDist = 1e6, ## define cis as < 1Mb
    useModel = modelLINEAR, ## test using linear models 
    verbose = TRUE,
    pvalue.hist = TRUE,
    min.pv.by.genesnp = FALSE,
    noFDRsaveMemory = FALSE)
```

This takes about 30 seconds.

### 4a. Aftermath: Number of associations

Here is the exact number of tests that were performed:
```{r}
eqtl$cis$ntests
```

> **Question 4:**: Above we guessed 1 test per SNP.
> How many tests were performed on average for each SNP?
> How could we calculate a Bonferroni-adjusted p-value 
> threshold for this number of tests?
```{r question4,include=show.answers}
```

At this threshold ($p < 1.03e-08$), we identify hundreds of associations.
```{r}
eqtls <- eqtl$cis$eqtls
eqtls <- eqtls[which(eqtls$pvalue < 1.03e-08),]
nrow(eqtls)
```

> **Question 5:**: How many eQTLs did we identify? 
> Careful!  It's not equal to `nrow(eqtls)`.
```{r question5,include=show.answers}
```

### 4b. Aftermath: Top associations

We will plot the strongest association:
```{r}
idx <- which.min(eqtls$pvalue)
eqtls[idx,]
```

We save the corresponding gene and SNP:
```{r}
top.gene <- as.character(eqtls$gene[idx])
top.snp <- as.character(eqtls$snps[idx])
top.gene
```
```{r}
top.snp
```

### 4c. Aftermath: Association plotting (data extraction)

To plot the association, we will need to extract the SNP and gene
expression values for the pair.  This can be a bit painful using
`SlicedData` objects because they don't support direct access to
individual rows or columns.  Here we create our own function to do
this.

It's a little complicated, but don't worry if you don't 
understand how it works, you just need to be able to use it!
```{r}
sliced.data.row <- function(object, row) {
    if (is.factor(row))
        row <- as.character(row)
    if (is.character(row)) {
        row <- which(rownames(object) == row)
        if (length(row) == 0)
            stop("Invalid row")
    }
    slice.size <- nrow(object[[1]])
    slice.idx <- floor(row/slice.size) + 1
    slice <- object[[slice.idx]]
    row <- row - slice.size * (slice.idx - 1)
    ret <- slice[row,]
    names(ret) <- colnames(object)
    ret             
}
```

### 4c. Aftermath: Association plotting

We can use this function to obtain the genotypes for any given SNP.
For example, below we extract data for the top SNP-gene pair:
```{r}
genotypes <- sliced.data.row(snp.data, top.snp)
genotypes <- factor(genotypes, levels=0:2, labels=c("AA","AB","BB"))
expression.levels <- sliced.data.row(rna.data, top.gene)
```

We plot their association as follows:
```{r}
boxplot(expression.levels ~ genotypes,
        main=paste(top.snp,top.gene),
        ylim=range(expression.levels),
        outline=F)
stripchart(expression.levels ~ genotypes,
           method="jitter", add=TRUE, vertical=TRUE,
           col=c("blue","orange","black"), pch=19)
```

### 5. (Optional) meQTL analysis

The code below will run the meQTL analysis:
```{r}
cpg.data <- SlicedData$new()
cpg.data$fileDelimiter <- ","           ## columns are separated by commas
cpg.data$fileOmitCharacters <- "NA"     ## missing values are represented by "NA"
cpg.data$fileSkipRows <- 1              ## the first column provides row names
cpg.data$fileSkipColumns <- 1           ## the first row provides column names
cpg.data$fileSliceSize <- 2000          ## only load the data 2000 rows at a time
cpg.data$LoadFile("data/cpg-data.csv") 

cpg.loc <- read.csv("data/cpg-features.csv", row.names=1)

threshold <- 0.05/nrow(snp.data)

meqtl <- Matrix_eQTL_main(
    snps = snp.data,
    gene = cpg.data,
    cvrt = covariates.data,
    pvOutputThreshold = 0,  ## ignore trans associations
    pvOutputThreshold.cis = threshold,
    output_file_name.cis = file.path(out.dir, "meqtl-cis.txt"),
    snpspos = snp.loc, 
    genepos = cpg.loc[,c("geneid","chr","left","right")],
    cisDist = 1e6,           ## define cis as < 1Mb
    useModel = modelLINEAR,
    verbose = TRUE,
    pvalue.hist = TRUE,
    min.pv.by.genesnp = FALSE,
    noFDRsaveMemory = FALSE)

threshold <- 0.05/meqtl$cis$ntests
meqtls <- meqtl$cis$eqtls
meqtls <- meqtls[which(meqtls$pvalue < threshold),]
```

### 6. (Optional) Associated trios

Such trios are of interest because they could indicate cases where a
SNP modifies methylation levels which in turn modify gene expression
levels.

Construct the trios:
```{r}
trios <- merge(
    data.frame(
        snp=eqtls$snps,
        gene=eqtls$gene,
        eqtl.statistic=eqtls$statistic),
    data.frame(
        snp=meqtls$snps,
        cpg=meqtls$gene,
        meqtl.statistic=meqtls$statistic))
```

There are 168 trios.

Calculate p-values for each gene/CpG site association:
```{r}
trios$p.value <- sapply(1:nrow(trios), function(i) {
    expr <- sliced.data.row(rna.data, trios$gene[i])
    meth <- sliced.data.row(cpg.data, trios$cpg[i])
    cor.test(meth, expr)$p.value
})
```

Identify the strongest gene/CpG site association.
```{r}
idx <- which.min(trios$p.value)
trios[idx,]
```
